src/tiled/scriptfileformatwrappers.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * scriptfileformatwrappers.h
+ * Copyright 2019, Thorbjørn Lindeijer <bjorn@lindeijer.nl>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "scriptfileformatwrappers.h"
+
+#include "editablemap.h"
+#include "editabletileset.h"
+#include "mapformat.h"
+#include "scriptedfileformat.h"
+#include "scriptmanager.h"
+#include "tilesetformat.h"
+
+#include <QCoreApplication>
+#include <QQmlEngine>
+
+namespace Tiled {
+
+ScriptTilesetFormatWrapper::ScriptTilesetFormatWrapper(TilesetFormat* format, QObject *parent)
+    : QObject(parent)
+    , mFormat(format)
+{}
+
+QString ScriptTilesetFormatWrapper::name() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedTilesetFormat*>(mFormat))
+        return scriptedFormat->name();
+    else
+        return mFormat->shortName();
+}
+
+QString ScriptTilesetFormatWrapper::extension() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedTilesetFormat*>(mFormat))
+        return scriptedFormat->extension();
+    else
+        return QString();
+}
+
+EditableTileset *ScriptTilesetFormatWrapper::read(QString &filename)
+{
+    auto tileset = mFormat->read(filename);
+    auto editable = new EditableTileset(tileset.data());
+    QQmlEngine::setObjectOwnership(editable, QQmlEngine::JavaScriptOwnership);
+    return editable;
+}
+
+QString ScriptTilesetFormatWrapper::write(EditableTileset *editable, QString &filename)
+{
+    auto tileset = editable->tileset();
+    if (mFormat->write(*tileset, filename))
+        return QString();
+    else
+        return mFormat->errorString();
+}
+
+bool ScriptTilesetFormatWrapper::supportsFile(QString &filename) const
+{
+    return mFormat->supportsFile(filename);
+}
+
+ScriptMapFormatWrapper::ScriptMapFormatWrapper()
+{
+    auto message = QCoreApplication::translate("Script Errors",
+                                               "MapFormatWrapper cannot be constructed");
+    ScriptManager::instance().throwError(message);
+}
+
+ScriptMapFormatWrapper::ScriptMapFormatWrapper(MapFormat *format, QObject *parent)
+    : QObject(parent)
+    , mFormat(format)
+{}
+
+QString ScriptMapFormatWrapper::name() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedMapFormat*>(mFormat))
+        return scriptedFormat->name();
+    else
+        return mFormat->shortName();
```

Comment:
Why not call this `shortName` and return `mFormat->shortName()` in both cases?

I'm not sure we should introduce the `name` and `extension` as properties. I understand I set them up that way for the scripted file formats, but actually they were just used to make the `nameFilter`. Do you need these properties somewhere? In hindsight I should probably just have made "nameFilter" a property of the scripted file format object instead of "name" and "extension", also because that actually gives more flexibility since it supports mentioning multiple extensions (`ScriptedFileFormat::supportsFile` could then be implemented using `Utils::fileNameMatchesNameFilter`).

Reply:
I wanted to include them because that makes the wrapper able to access everything in the original object that the user provides. It feels weird to accept a scripted object, and then not have access to the same fields as that object.

I can definitely do without them, just wanted to try to keep things consistent.

Reply:
Another option is to make the properties dynamic; they'd only exist if the backing TilesetFormat is castable to ScriptedTilesetFormat (similarly for MapFormat).

Reply:
I would prefer to do without any special handling for scripted map/tileset formats. And also to replace the `name` and `extension` properties with a `nameFilter` property, to match the `FileFormat` interface.